// containers/information/index.js
import React, {Component} from 'react'
import {connect} from 'react-redux'
import R from 'ramda'
import {Link} from 'react-router'

import {getInformationById} from 'selectors'
import {fetchInformationById, addInformationToBasket} from 'actions'
import BasketCart from 'components/basketCart'

class Information extends Component {
  componentDidMount () {
    this.props.fetchInformationById(this.props.params.id)
  }

  renderFields () {
    const {information} = this.props;
    const columnFields = R.compose(
      R.toPairs,
      R.pick([
        'cpu',
        'camera',
        'size',
        'weight',
        'display',
        'battery',
        'memory'
      ])
    )(information);

    return columnFields.map(([key, value]) => (
      <div className='column' key={key}>
        <div className='ab-details-title'>
          <p>{key}</p>
        </div>
        <div className='ab-details-info'>
          {value}
        </div>
      </div>
    ))
  }

  renderContent () {
    const {information} = this.props;

    return (
      <div className='thumbnail'>
        <div className='row'>
          <div className='col-md-6'>
            <img
              className='img-thumbnail'
              src={information.image}
              alt={information.name}
            />
          </div>
          <div className='col-md-6'>
            {this.renderFields()}
          </div>
        </div>
        <div className='caption-full'>
          <h4 className='pull-right'>${information.price}</h4>
          <h4>{information.name}</h4>
          <p>{information.description}</p>
        </div>
      </div>
    )
  }

  renderSidebar () {
    const {information, addInformationToBasket} = this.props;
    return (
      <div>
        <p className='lead'>Поиск</p>
        <BasketCart />
        <div className='form-group'>
          <h1>{information.name}</h1>
          <h2>${information.price}</h2>
        </div>
        <Link to='/' className='btn btn-info btn-block'>Вернуться</Link>
        <button
          type='button'
          className='btn btn-success btn-block'
          onClick={() => addInformationToBasket(information.id)}
        >
          Добавить в корзину
        </button>
      </div>
    )
  }

  render () {
    const {information} = this.props
    return (
      <div className='view-container'>
        <div className='container'>
          <div className='row'>
            <div className='col-md-9'>
              {information && this.renderContent()}
            </div>
            <div className='col-md-3'>
              {information && this.renderSidebar()}
            </div>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = state => {
  return {
    information: getInformationById(state, state.informationPage.id)
  }
};

const mapDispatchToProps = {
  fetchInformationById,
  addInformationToBasket
};

export default connect(mapStateToProps, mapDispatchToProps)(Information)
